wrappers/wrapper.rb

```diff
@@ -684,70 +602,120 @@ def detect_inconsistent_relation_timewindows_od_service(vrp, unfeasible, service
 
-      true
-    end
+      implicit_timewindow = [Models::Timewindow.new(start: 0)] # need to check time feasibility
 
-    def check_distances(vrp, unfeasible)
-      unfeasible = check(vrp, :time, unfeasible)
-      unfeasible = check(vrp, :distance, unfeasible)
-      unfeasible = check(vrp, :value, unfeasible)
+      vrp.vehicles.each{ |vehicle|
+        next unless service.vehicle_compatibility[vehicle.id] # already eliminated
 
-      vrp.services.each{ |service|
-        no_vehicle_compatible =
-          vrp.vehicles.none?{ |vehicle|
-            (service.activity ? [service.activity] : service.activities).any?{ |activity|
-              (service.skills.empty? || vehicle.skills.any?{ |skill_set| (service.skills - skill_set).empty? }) &&
-                (service.sticky_vehicle_ids.empty? || service.sticky_vehicle_ids.include?(vehicle.id)) &&
-                service_reachable_by_vehicle_within_timewindows(vrp, activity, vehicle)
-            }
+        vehicle_timewindows = [vehicle.timewindow, vehicle.sequence_timewindows].compact.flatten
+
+        next if !vrp.schedule? && vehicle.duration.nil? && vehicle.distance.nil? &&
+                s_activities.all?{ |a| a.timewindows.none?(&:end) } && vehicle_timewindows.none?(&:end)
+
+        vehicle_timewindows = implicit_timewindow if vehicle_timewindows.empty?
+
+        service.vehicle_compatibility[vehicle.id] =
+          s_activities.any?{ |activity|
+            time_to_go, time_to_return, dist_to_go, dist_to_return = two_way_time_and_dist(vrp, vehicle, activity) || [0, 0, 0, 0]
+
+            # NOTE: There is no easy way to include the setup duration in the elimination because the
+            # setup_duration logic is based of time[point_a][point_b] == 0; so we need to check all
+            # services which are 0 distance and then find the minimum (setup_duration + duration) and
+            # use this as the setup duration in the check below if it is less than the service.setup_duration
+
+            (vehicle.distance.nil? || dist_to_go + dist_to_return <= vehicle.distance) &&
+              (
+                vehicle.duration.nil? ||
+                time_to_go + vehicle.additional_service + vehicle.coef_service * activity.duration + time_to_return <= vehicle.duration
+              ) && (
+                (activity.timewindows.empty? ? implicit_timewindow : activity.timewindows).any?{ |s_tw|
+                  vehicle_timewindows.any?{ |v_tw|
+                    # vehicle has a tw that can serve service in time (incl. travel if it exists)
+                    (s_tw.day_index.nil? || v_tw.day_index.nil? || s_tw.day_index == v_tw.day_index) &&
+                      v_tw.start + time_to_go <= s_tw.safe_end(activity.late_multiplier&.positive?) &&
+                      [s_tw.start, v_tw.start + time_to_go].max + activity.duration.to_i + time_to_return <= v_tw.safe_end(vehicle.cost_late_multiplier&.positive?) &&
```

Comment:
I feel like we do need to extract this logic to a function but it needs both service and vehicle... 

We can put it 
:1st_place_medal:  under service with a vehicle as an argument
:2nd_place_medal: under vrp with both service and vehicle as arguments
:3rd_place_medal: under vehicle with service as an argument
:1234: All of the above :thinking: ?

Reply:
I would suggest to put it into activity with a vehicle as argument :wink:

Reply:
I let go the "over-writing the function" idea and went with the `x_on(vehicle)` and only implemented for duration and setup_duration because (these were already used) but total_duration depends on the previous service etc, so we can't really calculate it (and there is no use case at the moment).

Reply:
I totally agree with the `x_on(vehicle)` methods